DOCS-2483: Create write code quickstart (#3156)
viamrobotics/docs: docs/get-started/quickstarts/write-code.md

---
title: "Control machines with code in 4 minutes"

> linkTitle: "Program machines (4 min)"

Author: maybe not quite specific enough, let's go with
```suggestion
linkTitle: "Write code for machines (4 min)"
```

type: "docs"
weight: 80
cost: 75
images: ["/general/code.png"]
description: "Use Viam to gather images or sensor data from your machine."
---

Follow this guide to start writing code to perform actions with the {{< glossary_tooltip term_id="component" text="components" >}} or {{< glossary_tooltip term_id="service" text="services" >}} of a {{< glossary_tooltip term_id="machine" text="machine" >}}.

> {{<youtube embed_url="https://www.youtube-nocookie.com/embed/T9NxL-GJA9E">}}

Author: I see it starting immediately. I think what may have happened is I trimmed this in youtube and you saw an older version. Either way should be good

## Requirements

- A computer with a supported OS installed: Linux 64-bit, macOS, Windows Subsystem for Linux (WSL)
- A physical component such as a motor or a camera that is connected to the computer

> {{% expand "No computer or motor at hand?" %}}

Author: ```suggestion
{{% expand "No computer you can run `viam-server` on?" %}}
```

No problem.
Use [Try Viam](https://app.viam.com/try) to rent a rover online which is already configured with some components to test with.
If you are using a Try Viam rover **start with Step 4**.
{{% /expand%}}

## Instructions

Follow these steps to configure your machine inside the Viam app and write code to control it:

{{< expand "Step 1: Create a machine" >}}

Go to the Viam app.
Select a location and [add a new machine](/cloud/machines/#add-a-new-machine).

> ![A 'My Desk' location on the Viam app with a new machine name in the New machine field and the Add machine button next to the field highlighted.](/get-started/quickstarts/add-machine.png)

Author: good idea

> ![A 'My Desk' location on the Viam app with a new machine name in the New machine field and the Add machine button next to the field highlighted.](/get-started/quickstarts/add-machine.png)

Author: ```suggestion
![The 'First Location' page on the Viam app with a new machine name in the New machine field and the Add machine button next to the field highlighted.](/fleet/app-usage/create-machine.png)
```

{{< /expand >}}
{{< expand "Step 2: Install viam-server" >}}

Navigate to the **CONFIGURE** tab of your machine's page in the [Viam app](https://app.viam.com).
Follow the {{< glossary_tooltip term_id="setup" text="setup instructions" >}} that appear on your new machine's **CONFIGURE** page to install `viam-server` on your computer and connect it to the Viam app.

You can install `viam-server` on your personal computer, or on a single-board computer (SBC).
If you are using a microcontroller instead of a 64-bit computer, you can install a [lightweight version of `viam-server`](/get-started/installation/microcontrollers/).

{{< /expand >}}
{{< expand "Step 3: Configure a component or service">}}

Viam uses a composable system of building blocks called {{< glossary_tooltip term_id="resource" text="resources" >}} which you can combine according to your specific needs.
Components drive your hardware, and services provide high-level software functionality.

Click on the **+** icon to create a component or service.
If you are not sure what to add and your computer has a webcam attached, add a [`webcam`](/components/camera/webcam/).

These are the available components and services:

> <div class="cards max-page"><div class="row"><div class="col sectionlist"><div><h3>Components:</h3><ul class="sectionlist"><li><a href="../../../components/arm/" title="Arm Component"><div><picture><img src="/icons/components/arm.svg" width="/icons/components/arm" height="/icons/components/arm" alt="Arm" loading="lazy"></picture><p>Arm</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/base/" title="Base Component"><div><picture><img src="/icons/components/base.svg" width="/icons/components/base" height="/icons/components/base" alt="Base" loading="lazy"></picture><p>Base</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/board/" title="Board Component"><div><picture><img src="/icons/components/board.svg" width="/icons/components/board" height="/icons/components/board" alt="Board" loading="lazy"></picture><p>Board</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/camera/" title="Camera Component"><div><picture><img src="/icons/components/camera.svg" width="/icons/components/camera" height="/icons/components/camera" alt="Camera" loading="lazy"></picture><p>Camera</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/encoder/" title="Encoder Component"><div><picture><img src="/icons/components/encoder.svg" width="/icons/components/encoder" height="/icons/components/encoder" alt="Encoder" loading="lazy"></picture><p>Encoder</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/gantry/" title="Gantry Component"><div><picture><img src="/icons/components/gantry.svg" width="/icons/components/gantry" height="/icons/components/gantry" alt="Gantry" loading="lazy"></picture><p>Gantry</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/generic/" title="Generic Component"><div><picture><img src="/icons/components/generic.svg" width="/icons/components/generic" height="/icons/components/generic" alt="Generic" loading="lazy"></picture><p>Generic</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/gripper/" title="Gripper Component"><div><picture><img src="/icons/components/gripper.svg" width="/icons/components/gripper" height="/icons/components/gripper" alt="Gripper" loading="lazy"></picture><p>Gripper</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/input-controller/" title="Input Controller Component"><div><picture><img src="/icons/components/controller.svg" width="/icons/components/controller" height="/icons/components/controller" alt="Input Controller" loading="lazy"></picture><p>Input Controller</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/motor/" title="Motor Component"><div><picture><img src="/icons/components/motor.svg" width="/icons/components/motor" height="/icons/components/motor" alt="Motor" loading="lazy"></picture><p>Motor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/movement-sensor/" title="Movement Sensor Component"><div><picture><img src="/icons/components/imu.svg" width="/icons/components/imu" height="/icons/components/imu" alt="Movement Sensor" loading="lazy"></picture><p>Movement Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/power-sensor/" title="Power Sensor Component"><div><picture><img src="/icons/components/power-sensor.svg" width="/icons/components/power-sensor" height="/icons/components/power-sensor" alt="Power Sensor" loading="lazy"></picture><p>Power Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/sensor/" title="Sensor Component"><div><picture><img src="/icons/components/sensor.svg" width="/icons/components/sensor" height="/icons/components/sensor" alt="Sensor" loading="lazy"></picture><p>Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/servo/" title="Servo Component"><div><picture><img src="/icons/components/servo.svg" width="/icons/components/servo" height="/icons/components/servo" alt="Servo" loading="lazy"></picture><p>Servo</p></div></a></li></ul></div></div><div class="col sectionlist"><div><h3>Services:</h3><ul class="sectionlist"><li><a href="../../../services/data/" title="Data Management Service"><div><picture><img src="../../../services/icons/data-capture.svg" alt="Data Management" loading="lazy"></picture><p>Data Management</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/motion/" title="Motion Service"><div><picture><img src="../../../services/icons/motion.svg" alt="Motion" loading="lazy"></picture><p>Motion</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/frame-system/" title="The Robot Frame System"><div><picture><img src="../../../services/icons/frame-system.svg" alt="Frame System" loading="lazy"></picture><p>Frame System</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/base-rc/" title="Base Remote Control Service"><div><picture><img src="../../../services/icons/base-rc.svg" alt="Base Remote Control" loading="lazy"></picture><p>Base Remote Control</p></div></a></li></ul><ul class="sectionlist"><li><a href="/services/ml/" title="ML Model Service"><div><picture><img src="../../../services/icons/ml.svg" alt="ML Model" loading="lazy"></picture><p>ML Model</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/navigation/" title="The Navigation Service"><div><picture><img src="../../../services/icons/navigation.svg" alt="Navigation" loading="lazy"></picture><p>Navigation</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/slam/" title="SLAM Service"><div><picture><img src="../../../services/icons/slam.svg" alt="SLAM" loading="lazy"></picture><p>SLAM</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/vision/" title="Vision Service"><div><picture><img src="../../../services/icons/vision.svg" alt="Vision" loading="lazy"></picture><p>Vision</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/generic/" title="Generic Service"><div><picture><img src="../../../icons/components/generic.svg" alt="Generic" loading="lazy"></picture><p>Generic</p></div></a></li></ul></div></div></div></div>

Author: They would be but that's not ideal

> <div class="cards max-page"><div class="row"><div class="col sectionlist"><div><h3>Components:</h3><ul class="sectionlist"><li><a href="../../../components/arm/" title="Arm Component"><div><picture><img src="/icons/components/arm.svg" width="/icons/components/arm" height="/icons/components/arm" alt="Arm" loading="lazy"></picture><p>Arm</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/base/" title="Base Component"><div><picture><img src="/icons/components/base.svg" width="/icons/components/base" height="/icons/components/base" alt="Base" loading="lazy"></picture><p>Base</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/board/" title="Board Component"><div><picture><img src="/icons/components/board.svg" width="/icons/components/board" height="/icons/components/board" alt="Board" loading="lazy"></picture><p>Board</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/camera/" title="Camera Component"><div><picture><img src="/icons/components/camera.svg" width="/icons/components/camera" height="/icons/components/camera" alt="Camera" loading="lazy"></picture><p>Camera</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/encoder/" title="Encoder Component"><div><picture><img src="/icons/components/encoder.svg" width="/icons/components/encoder" height="/icons/components/encoder" alt="Encoder" loading="lazy"></picture><p>Encoder</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/gantry/" title="Gantry Component"><div><picture><img src="/icons/components/gantry.svg" width="/icons/components/gantry" height="/icons/components/gantry" alt="Gantry" loading="lazy"></picture><p>Gantry</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/generic/" title="Generic Component"><div><picture><img src="/icons/components/generic.svg" width="/icons/components/generic" height="/icons/components/generic" alt="Generic" loading="lazy"></picture><p>Generic</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/gripper/" title="Gripper Component"><div><picture><img src="/icons/components/gripper.svg" width="/icons/components/gripper" height="/icons/components/gripper" alt="Gripper" loading="lazy"></picture><p>Gripper</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/input-controller/" title="Input Controller Component"><div><picture><img src="/icons/components/controller.svg" width="/icons/components/controller" height="/icons/components/controller" alt="Input Controller" loading="lazy"></picture><p>Input Controller</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/motor/" title="Motor Component"><div><picture><img src="/icons/components/motor.svg" width="/icons/components/motor" height="/icons/components/motor" alt="Motor" loading="lazy"></picture><p>Motor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/movement-sensor/" title="Movement Sensor Component"><div><picture><img src="/icons/components/imu.svg" width="/icons/components/imu" height="/icons/components/imu" alt="Movement Sensor" loading="lazy"></picture><p>Movement Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/power-sensor/" title="Power Sensor Component"><div><picture><img src="/icons/components/power-sensor.svg" width="/icons/components/power-sensor" height="/icons/components/power-sensor" alt="Power Sensor" loading="lazy"></picture><p>Power Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/sensor/" title="Sensor Component"><div><picture><img src="/icons/components/sensor.svg" width="/icons/components/sensor" height="/icons/components/sensor" alt="Sensor" loading="lazy"></picture><p>Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/servo/" title="Servo Component"><div><picture><img src="/icons/components/servo.svg" width="/icons/components/servo" height="/icons/components/servo" alt="Servo" loading="lazy"></picture><p>Servo</p></div></a></li></ul></div></div><div class="col sectionlist"><div><h3>Services:</h3><ul class="sectionlist"><li><a href="../../../services/data/" title="Data Management Service"><div><picture><img src="../../../services/icons/data-capture.svg" alt="Data Management" loading="lazy"></picture><p>Data Management</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/motion/" title="Motion Service"><div><picture><img src="../../../services/icons/motion.svg" alt="Motion" loading="lazy"></picture><p>Motion</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/frame-system/" title="The Robot Frame System"><div><picture><img src="../../../services/icons/frame-system.svg" alt="Frame System" loading="lazy"></picture><p>Frame System</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/base-rc/" title="Base Remote Control Service"><div><picture><img src="../../../services/icons/base-rc.svg" alt="Base Remote Control" loading="lazy"></picture><p>Base Remote Control</p></div></a></li></ul><ul class="sectionlist"><li><a href="/services/ml/" title="ML Model Service"><div><picture><img src="../../../services/icons/ml.svg" alt="ML Model" loading="lazy"></picture><p>ML Model</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/navigation/" title="The Navigation Service"><div><picture><img src="../../../services/icons/navigation.svg" alt="Navigation" loading="lazy"></picture><p>Navigation</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/slam/" title="SLAM Service"><div><picture><img src="../../../services/icons/slam.svg" alt="SLAM" loading="lazy"></picture><p>SLAM</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/vision/" title="Vision Service"><div><picture><img src="../../../services/icons/vision.svg" alt="Vision" loading="lazy"></picture><p>Vision</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/generic/" title="Generic Service"><div><picture><img src="../../../icons/components/generic.svg" alt="Generic" loading="lazy"></picture><p>Generic</p></div></a></li></ul></div></div></div></div>

Author: ```suggestion
<div class="cards max-page"><div class="row"><div class="col sectionlist"><div><h3>Components:</h3><ul class="sectionlist"><li><a href="../../../components/arm/" title="Arm Component"><div><picture><img src="../../../icons/components/arm.svg" width="../../../icons/components/arm" height="../../../icons/components/arm" alt="Arm" loading="lazy"></picture><p>Arm</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/base/" title="Base Component"><div><picture><img src="../../../icons/components/base.svg" width="../../../icons/components/base" height="../../../icons/components/base" alt="Base" loading="lazy"></picture><p>Base</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/board/" title="Board Component"><div><picture><img src="../../../icons/components/board.svg" width="../../../icons/components/board" height="../../../icons/components/board" alt="Board" loading="lazy"></picture><p>Board</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/camera/" title="Camera Component"><div><picture><img src="../../../icons/components/camera.svg" width="../../../icons/components/camera" height="../../../icons/components/camera" alt="Camera" loading="lazy"></picture><p>Camera</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/encoder/" title="Encoder Component"><div><picture><img src="../../../icons/components/encoder.svg" width="../../../icons/components/encoder" height="../../../icons/components/encoder" alt="Encoder" loading="lazy"></picture><p>Encoder</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/gantry/" title="Gantry Component"><div><picture><img src="../../../icons/components/gantry.svg" width="../../../icons/components/gantry" height="../../../icons/components/gantry" alt="Gantry" loading="lazy"></picture><p>Gantry</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/generic/" title="Generic Component"><div><picture><img src="../../../icons/components/generic.svg" width="../../../icons/components/generic" height="../../../icons/components/generic" alt="Generic" loading="lazy"></picture><p>Generic</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/gripper/" title="Gripper Component"><div><picture><img src="../../../icons/components/gripper.svg" width="../../../icons/components/gripper" height="../../../icons/components/gripper" alt="Gripper" loading="lazy"></picture><p>Gripper</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/input-controller/" title="Input Controller Component"><div><picture><img src="../../../icons/components/controller.svg" width="../../../icons/components/controller" height="../../../icons/components/controller" alt="Input Controller" loading="lazy"></picture><p>Input Controller</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/motor/" title="Motor Component"><div><picture><img src="../../../icons/components/motor.svg" width="../../../icons/components/motor" height="../../../icons/components/motor" alt="Motor" loading="lazy"></picture><p>Motor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/movement-sensor/" title="Movement Sensor Component"><div><picture><img src="../../../icons/components/imu.svg" width="../../../icons/components/imu" height="../../../icons/components/imu" alt="Movement Sensor" loading="lazy"></picture><p>Movement Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/power-sensor/" title="Power Sensor Component"><div><picture><img src="../../../icons/components/power-sensor.svg" width="../../../icons/components/power-sensor" height="../../../icons/components/power-sensor" alt="Power Sensor" loading="lazy"></picture><p>Power Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/sensor/" title="Sensor Component"><div><picture><img src="../../../icons/components/sensor.svg" width="../../../icons/components/sensor" height="../../../icons/components/sensor" alt="Sensor" loading="lazy"></picture><p>Sensor</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../components/servo/" title="Servo Component"><div><picture><img src="../../../icons/components/servo.svg" width="../../../icons/components/servo" height="../../../icons/components/servo" alt="Servo" loading="lazy"></picture><p>Servo</p></div></a></li></ul></div></div><div class="col sectionlist"><div><h3>Services:</h3><ul class="sectionlist"><li><a href="../../../services/data/" title="Data Management Service"><div><picture><img src="../../../services/icons/data-capture.svg" alt="Data Management" loading="lazy"></picture><p>Data Management</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/motion/" title="Motion Service"><div><picture><img src="../../../services/icons/motion.svg" alt="Motion" loading="lazy"></picture><p>Motion</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/frame-system/" title="The Robot Frame System"><div><picture><img src="../../../services/icons/frame-system.svg" alt="Frame System" loading="lazy"></picture><p>Frame System</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/base-rc/" title="Base Remote Control Service"><div><picture><img src="../../../services/icons/base-rc.svg" alt="Base Remote Control" loading="lazy"></picture><p>Base Remote Control</p></div></a></li></ul><ul class="sectionlist"><li><a href="/services/ml/" title="ML Model Service"><div><picture><img src="../../../services/icons/ml.svg" alt="ML Model" loading="lazy"></picture><p>ML Model</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/navigation/" title="The Navigation Service"><div><picture><img src="../../../services/icons/navigation.svg" alt="Navigation" loading="lazy"></picture><p>Navigation</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/slam/" title="SLAM Service"><div><picture><img src="../../../services/icons/slam.svg" alt="SLAM" loading="lazy"></picture><p>SLAM</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/vision/" title="Vision Service"><div><picture><img src="../../../services/icons/vision.svg" alt="Vision" loading="lazy"></picture><p>Vision</p></div></a></li></ul><ul class="sectionlist"><li><a href="../../../services/generic/" title="Generic Service"><div><picture><img src="../../../icons/components/generic.svg" alt="Generic" loading="lazy"></picture><p>Generic</p></div></a></li></ul></div></div></div></div>
```

If you are unsure how to configure your [component](/components/) or a [service](/services/), follow the instructions in the model's documentation to configure any required attributes.

For more information, see [configuration](/configure/).

{{< /expand >}}
{{< expand "Step 4: Copy the sample code" >}}

Viam's APIs are standardized across all models of a given component or service.
This means you can test and change hardware without changing code.

After configuring your resource, navigate to your machine's **CONNECT** tab.
Click on any of the listed languages and follow the instructions to install the SDK.

> To install your preferred Viam SDK on your Linux or macOS development machine or [single-board computer](/components/board/), run one of the following commands in your terminal:

Author: I'll create a follow up ticket. We should, but currently with python those instructions aren't complete. I'll make a follow up ticket for the SDK team too

> To install your preferred Viam SDK on your Linux or macOS development machine or [single-board computer](/components/board/), run one of the following commands in your terminal:

Author: created tickets
